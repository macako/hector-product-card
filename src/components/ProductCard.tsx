import React, { CSSProperties, createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    maxCount,
    isMaxCountReached,
    increaseBy,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ counter, maxCount, product, increaseBy }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          increaseBy,
          product,
          reset,
        })}
      </div>
    </Provider>
  );
};
