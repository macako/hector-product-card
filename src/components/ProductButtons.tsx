import React, { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import style from '../styles/styles.module.css';

export const ProductButtons = ({ className }: { className?: string }) => {
  const { increaseBy, counter, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div className={`${style.buttonsContainer} ${className}`}>
      <button className={style.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={style.countLabel}>{counter}</div>
      <button
        disabled={false}
        className={`${style.buttonAdd} ${isMaxReached() && style.disabled}`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};
