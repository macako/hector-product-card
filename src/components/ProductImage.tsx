import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import style from '../styles/styles.module.css';

interface Props {
  image?: string;
  className?: string;
}

export const ProductImage = ({ image, className }: Props) => {
  const { product } = useContext(ProductContext);
  let imageToShow: string;

  if (image) {
    imageToShow = image;
  } else if (product.image) {
    imageToShow = product.image;
  } else {
    imageToShow = noImage;
  }

  return (
    <img
      className={`${style.productImg} ${className}`}
      src={imageToShow}
      alt="Product Image"
    />
  );
};
