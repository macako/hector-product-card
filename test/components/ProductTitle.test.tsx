import React from 'react';
import render from 'react-test-renderer';

import { ProductTitle } from '../../src/components';

describe('<ProductTitle />', () => {
  test('should display component as expected', () => {
    const wrapper = render.create(<ProductTitle title="test" />);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
