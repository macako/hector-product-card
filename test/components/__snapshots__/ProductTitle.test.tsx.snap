// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTitle /> should display component as expected 1`] = `
<span
  className="productDescription undefined"
>
  test
</span>
`;
